In [3]:
import PIL
from PIL import Image
import pandas as pd
import io
import os
import math
import numpy as np
import itertools

import pyarrow.dataset as ds
import pyarrow.parquet as pq

import matplotlib.pyplot as plt
import matplotlib.patches as patches

import shutil
from google.colab import drive

In [ ]:
from google.colab import auth
auth.authenticate_user()

In [ ]:
cam_img_bucket_path = "gs://waymo_open_dataset_v_2_0_1/training/camera_image"
cam_box_bucket_path = "gs://waymo_open_dataset_v_2_0_1/training/camera_box"

In [ ]:
# create directories
os.makedirs(
    "waymo_open_dataset_v_2_0_1/training/front_camera_images",
    exist_ok=True
)
os.makedirs(
    "waymo_open_dataset_v_2_0_1/training/front_camera_box",
    exist_ok=True
)
if os.path.isdir("waymo_open_dataset_v_2_0_1/training/front_camera_images"):
  print(True)

True


In [ ]:
# import pyarrow.dataset as ds
# import pyarrow.parquet as pq
# import os
# import gc

# # GCS path

# image_path = "gs://waymo_open_dataset_v_2_0_1/training/camera_image"
# box_path = "gs://waymo_open_dataset_v_2_0_1/training/camera_box"

# output_dir = "waymo_open_dataset_v_2_0_1/training/front_camera"
# os.makedirs(output_dir, exist_ok=True)

# image_output = os.path.join(
#     output_dir,
#     "front_camera_images.parquet"
# )

# box_output = os.path.join(
#     output_dir,
#     "front_camera_boxes.parquet"
# )

# TARGET_IMAGES = 5000




# image_dataset = ds.dataset(
#     image_path,
#     format="parquet"
# )

# image_scanner = image_dataset.scanner(
#     columns=[
#         "key.segment_context_name",
#         "key.frame_timestamp_micros",
#         "key.camera_name",
#         "[CameraImageComponent].image"
#     ],
#     filter=ds.field("key.camera_name") == 1,
#     batch_size=64
# )

# image_writer = None
# total_images = 0

# for batch in image_scanner.to_batches():

#     remaining = TARGET_IMAGES - total_images

#     if batch.num_rows > remaining:
#         batch = batch.slice(0, remaining)

#     if image_writer is None:
#         image_writer = pq.ParquetWriter(
#             image_output,
#             batch.schema
#         )

#     image_writer.write_batch(batch)

#     total_images += batch.num_rows

#     print(f"Images extracted: {total_images}/{TARGET_IMAGES}")

#     del batch
#     gc.collect()

#     if total_images >= TARGET_IMAGES:
#         break

# if image_writer is not None:
#     image_writer.close()

# print("\nImage Extraction complete.")
# print(f"Images: {total_images}")


In [ ]:
# # images from 5000 - 10000
# import pyarrow.dataset as ds
# import pyarrow.parquet as pq
# import os
# import gc


# # Dataset

# cam_img_path = "gs://waymo_open_dataset_v_2_0_1/training/camera_image"

# img_dataset = ds.dataset(
#     cam_img_path,
#     format="parquet"
# )


# # Scanner


# scanner = img_dataset.scanner(
#     columns=[
#         "key.segment_context_name",
#         "key.frame_timestamp_micros",
#         "key.camera_name",
#         "[CameraImageComponent].image"
#     ],
#     filter=ds.field("key.camera_name") == 1,
#     batch_size=32
# )


# # Output


# output_dir = "/content/waymo_open_dataset_v_2_0_1/training/front_camera"

# os.makedirs(output_dir, exist_ok=True)

# output_path = os.path.join(
#     output_dir,
#     "front_camera_images_5000_10000.parquet"
# )


# # Parameters


# SKIP_IMAGES = 5000
# TARGET_IMAGES = 5000

# skipped = 0
# extracted = 0

# writer = None


# # Extract


# for batch in scanner.to_batches():


#     # Skip first 5,000 FRONT images


#     if skipped < SKIP_IMAGES:

#         remaining_to_skip = SKIP_IMAGES - skipped

#         if batch.num_rows <= remaining_to_skip:
#             skipped += batch.num_rows
#             continue

#         # Batch contains some images we need to skip
#         batch = batch.slice(
#             remaining_to_skip
#         )

#         skipped = SKIP_IMAGES


#     # Don't exceed next 5,000


#     remaining = TARGET_IMAGES - extracted

#     if batch.num_rows > remaining:
#         batch = batch.slice(
#             0,
#             remaining
#         )


#     # Create writer


#     if writer is None:
#         writer = pq.ParquetWriter(
#             output_path,
#             batch.schema
#         )


#     # Write batch


#     writer.write_batch(batch)

#     extracted += batch.num_rows

#     print(
#         f"Extracted: {extracted}/{TARGET_IMAGES}"
#     )

#     del batch
#     gc.collect()

#     if extracted >= TARGET_IMAGES:
#         break


# # Close


# if writer is not None:
#     writer.close()

# print(
#     f"\nDone. Extracted images "
#     f"{SKIP_IMAGES + 1} to "
#     f"{SKIP_IMAGES + extracted}."
# )

# print(f"Saved to: {output_path}")

In [ ]:

# # 2. Extract FRONT camera boxes


# box_dataset = ds.dataset(
#     cam_box_bucket_path,
#     format="parquet"
# )

# box_scanner = box_dataset.scanner(
#     filter=ds.field("key.camera_name") == 1,
#     batch_size=2048
# )

# box_writer = None
# total_boxes = 0

# for batch in box_scanner.to_batches():

#     if box_writer is None:
#         box_writer = pq.ParquetWriter(
#             box_output,
#             batch.schema
#         )

#     box_writer.write_batch(batch)

#     total_boxes += batch.num_rows

#     print(f"Boxes extracted: {total_boxes}")

#     del batch
#     gc.collect()

# if box_writer is not None:
#     box_writer.close()


# print("\nBox Extraction complete.")
# print(f"Boxes:  {total_boxes}")

In [ ]:
# import os

# output_dir = "/content/drive/MyDrive/waymo/front_camera"

# os.makedirs(output_dir, exist_ok=True)

In [ ]:
# image_output = os.path.join(
#     output_dir,
#     "front_camera_images.parquet"
# )

# box_output = os.path.join(
#     output_dir,
#     "front_camera_boxes.parquet"
# )

In [ ]:
# print("Directory:", output_dir)
# print("Exists:", os.path.exists(output_dir))
# print("Contents:", os.listdir(output_dir))

Directory: /content/drive/MyDrive/waymo/front_camera
Exists: True
Contents: []


In [ ]:
# from google.colab import files

# files.download(
#     "/content/waymo_open_dataset_v_2_0_1/training/front_camera/front_camera_images.parquet"
# )

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
# Downloading images at a timestamp skip of 2 sec to avoid non chaning frames.
import pyarrow.dataset as ds
import pyarrow.parquet as pq
import os
import gc


# Paths


image_path = "gs://waymo_open_dataset_v_2_0_1/training/camera_image"

output_dir = "./waymo_extracted"
os.makedirs(output_dir, exist_ok=True)

image_output = os.path.join(
    output_dir,
    "front_camera_images_2sec.parquet"
)

TARGET_IMAGES = 10000
GAP_MICROS = 2_000_000   # 2 seconds



# Dataset


image_dataset = ds.dataset(
    image_path,
    format="parquet"
)

image_scanner = image_dataset.scanner(
    columns=[
        "key.segment_context_name",
        "key.frame_timestamp_micros",
        "key.camera_name",
        "[CameraImageComponent].image"
    ],
    filter=ds.field("key.camera_name") == 1,
    batch_size=64
)



# Temporal sampling


# Last selected timestamp for each segment
last_timestamp = {}

image_writer = None
total_images = 0

for batch in image_scanner.to_batches():

    selected_indices = []

    segments = batch["key.segment_context_name"]
    timestamps = batch["key.frame_timestamp_micros"]

    for i in range(batch.num_rows):

        segment = segments[i].as_py()
        timestamp = timestamps[i].as_py()

        # First frame from this segment
        if segment not in last_timestamp:
            selected_indices.append(i)
            last_timestamp[segment] = timestamp

        # At least 2 seconds since previous selected frame
        elif timestamp - last_timestamp[segment] >= GAP_MICROS:
            selected_indices.append(i)
            last_timestamp[segment] = timestamp

        # Stop once we have enough
        if total_images + len(selected_indices) >= TARGET_IMAGES:
            break

    # Nothing selected from this batch
    if not selected_indices:
        continue

    # Keep only selected rows
    selected_batch = batch.take(selected_indices)

    # Don't exceed TARGET_IMAGES
    remaining = TARGET_IMAGES - total_images

    if selected_batch.num_rows > remaining:
        selected_batch = selected_batch.slice(0, remaining)

    # Initialize writer
    if image_writer is None:
        image_writer = pq.ParquetWriter(
            image_output,
            selected_batch.schema
        )

    image_writer.write_batch(selected_batch)

    total_images += selected_batch.num_rows

    print(
        f"Images extracted: "
        f"{total_images}/{TARGET_IMAGES}"
    )

    del selected_batch
    del batch
    gc.collect()

    if total_images >= TARGET_IMAGES:
        break



# Close


if image_writer is not None:
    image_writer.close()

print("\nExtraction complete.")
print(f"Images: {total_images}")
print(f"Saved: {image_output}")

In [ ]:
# The actual time span of your extracted data:
import pyarrow.parquet as pq

table = pq.read_table(
    "./waymo_extracted/front_camera_images_2sec.parquet",
    columns=[
        "key.segment_context_name",
        "key.frame_timestamp_micros"
    ]
)

print("Rows:", table.num_rows)

timestamps = table["key.frame_timestamp_micros"].to_pylist()

print("Min timestamp:", min(timestamps))
print("Max timestamp:", max(timestamps))

Rows: 7980
Min timestamp: 1506904088695574
Max timestamp: 1573312617324815


In [ ]:
# how many samples each segment contributed:
from collections import Counter

segments = table["key.segment_context_name"].to_pylist()

counts = Counter(segments)

print("Number of segments:", len(counts))
print("Largest segment:", max(counts.values()))
print("Smallest segment:", min(counts.values()))

Number of segments: 798
Largest segment: 10
Smallest segment: 10


In [ ]:
# Check the timestamps within one segment
import pyarrow.parquet as pq

table = pq.read_table(
    "./waymo_extracted/front_camera_images_2sec.parquet",
    columns=[
        "key.segment_context_name",
        "key.frame_timestamp_micros"
    ]
)

segments = table["key.segment_context_name"].to_pylist()
timestamps = table["key.frame_timestamp_micros"].to_pylist()

first_segment = segments[0]

for s, t in zip(segments, timestamps):
    if s == first_segment:
        print(t)

1550083467346370
1550083469445090
1550083471445769
1550083473448247
1550083475449087
1550083477548901
1550083479548969
1550083481648785
1550083483748615
1550083485748805


In [8]:
# Save extracted images from GLI to drive
# 1. Mount Google Drive if not already mounted
if not os.path.exists('/content/drive'):
    drive.mount('/content/drive')

src = "/content/waymo_extracted/front_camera_images_2sec.parquet"
dst_dir = "/content/drive/MyDrive/waymo"
dst = os.path.join(dst_dir, "front_camera_images_2sec.parquet")

# 2. Check if source file actually exists
if not os.path.exists(src):
    raise FileNotFoundError(f"Source file does not exist at: {src}")

# 3. Create destination directory if it doesn't exist
os.makedirs(dst_dir, exist_ok=True)

# 4. Copy file with size verification
print("Copying file to Google Drive... Please wait.")
shutil.copy2(src, dst)

# 5. Verify the file was written
if os.path.exists(dst):
    size_mb = os.path.getsize(dst) / (1024 * 1024)
    print(f"Success! File saved to: {dst} ({size_mb:.2f} MB)")
else:
    print("Copy failed: File was not created in Google Drive.")

Mounted at /content/drive
Copying file to Google Drive... Please wait.
Success! File saved to: /content/drive/MyDrive/waymo/front_camera_images_5000_10000.parquet (3099.38 MB)


In [4]:
# keys to match the image data to box data
keys = [
    "key.segment_context_name",
    "key.frame_timestamp_micros",
    "key.camera_name",
]
# Extract the image and box data saved in colab runtime disk into pyarrow dataset
box_dataset = ds.dataset("/content/waymo_extracted/front_camera_boxes.parquet")
box_dataset.schema

img_dataset = ds.dataset("/content/waymo_extracted/front_camera_images_2sec.parquet")
img_dataset.schema

key.segment_context_name: string
key.frame_timestamp_micros: int64
key.camera_name: int8
[CameraImageComponent].image: binary
-- schema metadata --
pandas: '{"column_indexes": [{"field_name": null, "metadata": {"encoding"' + 3109

In [5]:
import os
# avaible CPU thread pool in google colab
print(f"Available CPU threads: {os.cpu_count()}")

Available CPU threads: 2


In [6]:
import pyarrow
# Pyarrow thread pool
pyarrow.cpu_count()

2

Note that the exact CPU thread count in google colab depends on whether you are on free tier or google colab/pro+ version. Here there are only 2 cpu threads are present.

Important :- Directly inner joining image and box pyarrow dataset will require around 12 - 16 GB RAM and using multithreading (`use_thread=True`) will require faster CPU execution which results in peak RAM consumption, all these reasons will definately result in colab runtime crash. So to avoid it, i created an batch scanner for box data, put the 7980 images keys into python set and then matched all the keys batch by batch box data which took some time to compute and required lesser RAM.

In [ ]:
# join data (avoid this if your RAM<16GB)
seg_data = img_dataset.join(right_dataset=box_dataset, keys=keys, join_type="inner", use_threads=True)

- `pq.read_table()` will load the full data for all specified columns into RAM.

- PyArrow will read every raw compressed `JPEG` byte blob into system memory and construct an in-memory `pyarrow.Table` .

In [9]:
# create an pyarrow table (This will load all the data into the RAM )
import pyarrow.parquet as pq

img_table = pq.read_table(
    "/content/waymo_extracted/front_camera_images_2sec.parquet",
    columns=[
        "key.segment_context_name",
        "key.frame_timestamp_micros",
        "key.camera_name",
        "[CameraImageComponent].image"
    ]
)

keys = [
    "key.segment_context_name",
    "key.frame_timestamp_micros",
    "key.camera_name"
]

In [10]:
# compress image keys data
image_keys = set(zip(
    img_table["key.segment_context_name"].to_pylist(),
    img_table["key.frame_timestamp_micros"].to_pylist(),
    img_table["key.camera_name"].to_pylist()
))

print(len(image_keys))

7980


In [17]:
# Create an scanner on box image data
box_dataset = ds.dataset(
    "/content/waymo_extracted/front_camera_boxes.parquet",
    format="parquet"
)

box_scanner = box_dataset.scanner(
    columns=[
        "key.segment_context_name",
        "key.frame_timestamp_micros",
        "key.camera_name",
        "key.camera_object_id",
        "[CameraBoxComponent].box.center.x",
        "[CameraBoxComponent].box.center.y",
        "[CameraBoxComponent].box.size.x",
        "[CameraBoxComponent].box.size.y",
        "[CameraBoxComponent].type",
        "[CameraBoxComponent].difficulty_level.detection"
    ],
    batch_size=16384
)

In [ ]:
# Map the compressed image key data batch-wise in box scanner, and save the data into colab disk
import pyarrow as pa
import gc

box_writer = None
total_boxes = 0

for batch in box_scanner.to_batches():

    segments = batch["key.segment_context_name"].to_pylist()
    timestamps = batch["key.frame_timestamp_micros"].to_pylist()
    cameras = batch["key.camera_name"].to_pylist()

    # Find rows belonging to our selected images
    indices = [
        i
        for i, key in enumerate(zip(segments, timestamps, cameras))
        if key in image_keys
    ]

    if indices:

        matched_batch = batch.take(pa.array(indices))

        if box_writer is None:
            box_writer = pq.ParquetWriter(
                "front_camera_boxes_matched.parquet",
                matched_batch.schema
            )

        box_writer.write_batch(matched_batch)

        total_boxes += matched_batch.num_rows

        print(f"Matched boxes: {total_boxes:,}")

        del matched_batch

    del batch
    gc.collect()



# Close writer


if box_writer is not None:
    box_writer.close()

print("\nBox extraction complete.")
print(f"Total matched boxes: {total_boxes:,}")

In [19]:
# check the saved matched box data schema (it should match original box data schema)
data = pq.ParquetFile("/content/front_camera_boxes_matched.parquet")
data.schema

required group field_id=-1 schema {
  optional binary field_id=-1 key.segment_context_name (String);
  optional int64 field_id=-1 key.frame_timestamp_micros;
  optional int32 field_id=-1 key.camera_name (Int(bitWidth=8, isSigned=true));
  optional binary field_id=-1 key.camera_object_id (String);
  optional double field_id=-1 [CameraBoxComponent].box.center.x;
  optional double field_id=-1 [CameraBoxComponent].box.center.y;
  optional double field_id=-1 [CameraBoxComponent].box.size.x;
  optional double field_id=-1 [CameraBoxComponent].box.size.y;
  optional int32 field_id=-1 [CameraBoxComponent].type (Int(bitWidth=8, isSigned=true));
  optional int32 field_id=-1 [CameraBoxComponent].difficulty_level.detection (Int(bitWidth=8, isSigned=true));
}

In [20]:
data.metadata

  created_by: parquet-cpp-arrow version 23.0.1
  num_columns: 10
  num_rows: 184880
  num_row_groups: 3754
  format_version: 2.6
  serialized_size: 5184611

In [24]:
# save matched box data into google drive
# 1. Mount Google Drive if not already mounted
if not os.path.exists('/content/drive'):
    drive.mount('/content/drive')

src = "/content/front_camera_boxes_matched.parquet"
dst_dir = "/content/drive/MyDrive/waymo"
dst = os.path.join(dst_dir, "front_camera_boxes_matched.parquet")

# 2. Check if source file actually exists
if not os.path.exists(src):
    raise FileNotFoundError(f"Source file does not exist at: {src}")

# 3. Create destination directory if it doesn't exist
os.makedirs(dst_dir, exist_ok=True)

# 4. Copy file with size verification
print("Copying file to Google Drive... Please wait.")
shutil.copy2(src, dst)

# 5. Verify the file was written
if os.path.exists(dst):
    size_mb = os.path.getsize(dst) / (1024 * 1024)
    print(f"Success! File saved to: {dst} ({size_mb:.2f} MB)")
else:
    print("Copy failed: File was not created in Google Drive.")

Copying file to Google Drive... Please wait.
Success! File saved to: /content/drive/MyDrive/waymo/front_camera_boxes_matched.parquet (18.64 MB)
